File: Ios/BarkCloud/BarkCloud/Networking/FileTransferService.swift
import Foundation
import GRPCCore

enum FileTransferError: Error {
    case badURL
    case badUploadResponse
    case downloadFailed
}

/// Передача байтов файлов: gRPC `FilesApi` (получение ссылок/квоты) + обычный HTTP
/// upload/download на готовые URL, которые возвращает сервер. Загрузка/скачивание
/// идут НЕ через gRPC, а POST/GET на `:7025/web/upload|download/{id}` через
/// `InsecureHTTP.session` (self-signed TLS).
final class FileTransferService: Sendable {
    private let grpc: GrpcManager

    init(grpc: GrpcManager) {
        self.grpc = grpc
    }

    // MARK: - gRPC (FilesApi)

    /// Получить адрес для загрузки и предварительный file_id.
    func getUploadURL(type: Barkcloud_Files_UploadFileType) async throws -> (url: String, fileID: String) {
        let stub = try await grpc.filesStub()
        var req = Barkcloud_Files_GetUploadUrlRequest()
        req.fileType = type
        let resp = try await stub.getUploadUrl(req)
        return (resp.url, resp.fileID)
    }

    /// Временные ссылки на оригиналы по file_id (file_id → URL).
    func tempDownloadURLs(fileIDs: [String]) async throws -> [String: URL] {
        guard !fileIDs.isEmpty else { return [:] }
        let stub = try await grpc.filesStub()
        var req = Barkcloud_Files_GetTempDownloadUrlRequest()
        req.fileIds = fileIDs
        let resp = try await stub.getTempDownloadUrl(req)
        var result: [String: URL] = [:]
        for item in resp.fileUrls where !item.url.isEmpty {
            result[item.fileID] = URL(string: item.url)
        }
        return result
    }

    /// Информация о хранилище пользователя (использовано / лимит, в байтах).
    func storageInfo() async throws -> (used: Int64, limit: Int64) {
        let stub = try await grpc.filesStub()
        let resp = try await stub.getUserStorageInfo(Barkcloud_Files_GetUserStorageInfoRequest())
        return (resp.totalUsedStorage, resp.storageLimit)
    }

    // MARK: - HTTP

    /// Залить байты по адресу из `getUploadURL`. Возвращает `fileId` ИЗ ОТВЕТА —
    /// при дедупликации он может отличаться от запрошенного, всегда используем его.
    func upload(data: Data, fileName: String, to urlString: String) async throws -> String {
        guard let url = URL(string: urlString) else { throw FileTransferError.badURL }

        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        if let token = await grpc.validAccessToken(), !token.isEmpty {
            request.setValue(token, forHTTPHeaderField: "x-auth-token")
        }
        let boundary = "Boundary-\(UUID().uuidString)"
        request.setValue("multipart/form-data; boundary=\(boundary)", forHTTPHeaderField: "Content-Type")
        let body = Self.multipartBody(boundary: boundary, fieldName: "file", fileName: fileName, data: data)

        let (respData, response) = try await InsecureHTTP.session.upload(for: request, from: body)
        guard let http = response as? HTTPURLResponse, (200..<300).contains(http.statusCode) else {
            throw FileTransferError.badUploadResponse
        }
        guard let obj = try? JSONSerialization.jsonObject(with: respData) as? [String: Any],
              let fileID = obj["fileId"] as? String, !fileID.isEmpty else {
            throw FileTransferError.badUploadResponse
        }
        return fileID
    }

    /// Скачать оригинал во временный файл (для предпросмотра / шеринга).
    func download(from url: URL, suggestedName: String) async throws -> URL {
        let (tempURL, response) = try await InsecureHTTP.session.download(from: url)
        guard let http = response as? HTTPURLResponse, (200..<300).contains(http.statusCode) else {
            throw FileTransferError.downloadFailed
        }
        let name = suggestedName.isEmpty ? UUID().uuidString : suggestedName
        let dest = FileManager.default.temporaryDirectory.appendingPathComponent(name)
        try? FileManager.default.removeItem(at: dest)
        try FileManager.default.moveItem(at: tempURL, to: dest)
        return dest
    }

    private static func multipartBody(boundary: String, fieldName: String, fileName: String, data: Data) -> Data {
        var body = Data()
        body.append("--\(boundary)\r\n".data(using: .utf8)!)
        body.append("Content-Disposition: form-data; name=\"\(fieldName)\"; filename=\"\(fileName)\"\r\n".data(using: .utf8)!)
        body.append("Content-Type: application/octet-stream\r\n\r\n".data(using: .utf8)!)
        body.append(data)
        body.append("\r\n--\(boundary)--\r\n".data(using: .utf8)!)
        return body
    }
}

File: Android/BarkCloud.Android/app/src/main/java/com/barkfluff/BarkCloud/net/FileTransferService.kt
package com.barkfluff.BarkCloud.net

import android.content.Context
import android.net.Uri
import android.provider.OpenableColumns
import barkcloud.files.FilesApiOuterClass.GetTempDownloadUrlRequest
import barkcloud.files.FilesApiOuterClass.GetUploadUrlRequest
import barkcloud.files.FilesApiOuterClass.GetUserStorageInfoRequest
import barkcloud.files.FilesApiOuterClass.UploadFileType
import com.barkfluff.BarkCloud.data.GlobalParam
import com.barkfluff.BarkCloud.grpc.GrpcManager
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import okhttp3.MediaType.Companion.toMediaTypeOrNull
import okhttp3.MultipartBody
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.RequestBody
import okhttp3.RequestBody.Companion.toRequestBody
import okio.BufferedSink
import okio.source
import org.json.JSONObject
import java.io.File
import java.io.IOException
import java.util.UUID

/**
 * Передача байтов файлов: gRPC `FilesApi` (ссылки/квота) + обычный HTTP upload/download
 * на готовые URL, которые возвращает сервер. Загрузка/скачивание идут НЕ через gRPC, а
 * POST/GET на `:7025/web/{upload|download}/{id}` через [InsecureHttp] (self-signed TLS).
 */
class FileTransferService(
    private val appContext: Context,
    private val grpc: GrpcManager,
    private val globalParam: GlobalParam,
    private val http: OkHttpClient,
) {

    // MARK: gRPC (FilesApi)

    /** Получить адрес для загрузки и предварительный file_id. */
    suspend fun getUploadUrl(type: UploadFileType): UploadTarget {
        val resp = grpc.filesStub().getUploadUrl(
            GetUploadUrlRequest.newBuilder().setFileType(type).build()
        )
        return UploadTarget(resp.url, resp.fileId)
    }

    /** Временные ссылки на оригиналы по file_id (file_id → URL, только непустые). */
    suspend fun tempDownloadUrls(fileIds: List<String>): Map<String, String> {
        if (fileIds.isEmpty()) return emptyMap()
        val resp = grpc.filesStub().getTempDownloadUrl(
            GetTempDownloadUrlRequest.newBuilder().addAllFileIds(fileIds).build()
        )
        return resp.fileUrlsList
            .filter { it.url.isNotEmpty() }
            .associate { it.fileId to it.url }
    }

    /** Информация о хранилище (использовано / лимит, в байтах). */
    suspend fun storageInfo(): StorageInfo {
        val resp = grpc.filesStub().getUserStorageInfo(GetUserStorageInfoRequest.getDefaultInstance())
        return StorageInfo(resp.totalUsedStorage, resp.storageLimit)
    }

    // MARK: HTTP

    /** Залить содержимое [uri] (стримингом). Возвращает file_id ИЗ ОТВЕТА (учёт дедупликации). */
    suspend fun upload(uri: Uri, fileName: String, urlString: String): String =
        upload(uriRequestBody(uri), fileName, urlString)

    /** Залить локальный staged-файл из app storage. */
    suspend fun upload(file: File, fileName: String, urlString: String): String =
        upload(fileRequestBody(file), fileName, urlString)

    /** Залить готовые байты (например, аватар). */
    suspend fun upload(bytes: ByteArray, fileName: String, urlString: String): String =
        upload(bytes.toRequestBody(OCTET_STREAM), fileName, urlString)

    private suspend fun upload(body: RequestBody, fileName: String, urlString: String): String =
        withContext(Dispatchers.IO) {
            val multipart = MultipartBody.Builder()
                .setType(MultipartBody.FORM)
                .addFormDataPart("file", fileName, body)
                .build()
            val builder = Request.Builder().url(urlString).post(multipart)
            globalParam.accessToken?.takeIf { it.isNotBlank() }?.let { builder.header("x-auth-token", it) }
            http.newCall(builder.build()).execute().use { resp ->
                if (!resp.isSuccessful) throw IOException("Upload failed: HTTP ${resp.code}")
                val text = resp.body?.string().orEmpty()
                val fileId = runCatching { JSONObject(text).optString("fileId") }.getOrNull()
                if (fileId.isNullOrEmpty()) throw IOException("Upload: no fileId in response")
                fileId
            }
        }

    /** Скачать оригинал во временный файл (для предпросмотра / шеринга). */
    suspend fun download(urlString: String, suggestedName: String): File =
        withContext(Dispatchers.IO) {
            http.newCall(Request.Builder().url(urlString).get().build()).execute().use { resp ->
                if (!resp.isSuccessful) throw IOException("Download failed: HTTP ${resp.code}")
                val name = suggestedName.ifEmpty { UUID.randomUUID().toString() }
                val dest = File(appContext.cacheDir, name)
                if (dest.exists()) dest.delete()
                val source = resp.body ?: throw IOException("Download: empty body")
                source.byteStream().use { input -> dest.outputStream().use { input.copyTo(it) } }
                dest
            }
        }

    private fun uriRequestBody(uri: Uri): RequestBody = object : RequestBody() {
        override fun contentType() = OCTET_STREAM
        override fun contentLength(): Long = querySize(uri)
        override fun writeTo(sink: BufferedSink) {
            appContext.contentResolver.openInputStream(uri)?.use { input ->
                sink.writeAll(input.source())
            } ?: throw IOException("Cannot open $uri")
        }
    }

    private fun fileRequestBody(file: File): RequestBody = object : RequestBody() {
        override fun contentType() = OCTET_STREAM
        override fun contentLength(): Long = file.length()
        override fun writeTo(sink: BufferedSink) {
            file.inputStream().use { input -> sink.writeAll(input.source()) }
        }
    }

    private fun querySize(uri: Uri): Long = runCatching {
        appContext.contentResolver.query(uri, arrayOf(OpenableColumns.SIZE), null, null, null)?.use { c ->
            val idx = c.getColumnIndex(OpenableColumns.SIZE)
            if (c.moveToFirst() && idx >= 0 && !c.isNull(idx)) c.getLong(idx) else -1L
        } ?: -1L
    }.getOrDefault(-1L)

    data class UploadTarget(val url: String, val fileId: String)
    data class StorageInfo(val used: Long, val limit: Long)

    private companion object {
        val OCTET_STREAM = "application/octet-stream".toMediaTypeOrNull()
    }
}
